src/semck/typeck.rs

```diff
@@ -203,6 +203,16 @@ impl<'a, 'ast> TypeCheck<'a, 'ast> {
         }
     }
 
+    fn check_stmt_defer(&mut self, s: &'ast StmtDeferType) {
```

Comment:
`self.visit_expr(&s.expr)` is missing. This is necessary to type check the defer expression. Right now this method only checks if the given expression is a function call.